### PR TITLE
Add grpc based kaetzchen service plugin system

### PR DESCRIPTION
**Pull request by david415**
This feature addition adds a modular plugin system to
the Katzenpost mix server which allows you to write a mixnet
service in any language with a gRPC library.

I've recently cleaned up the code quite a bit but perhaps there are things that I've missed and that is why I am requesting code review. Please use the inline github code review thingy. Otherwise review rant on irc is fine, thanks.

### CODE REVIEW

**Comment by david415** [comment]
This feature addition uses the go-plugin library:
https://github.com/hashicorp/go-plugin

to implement a kaetzchen plugin system; that is external programs which
listen on a UNIX domain socket for connections from the katzenpost server. To better understand the nature of this change to the katzenpost server, stare at this computation pipeline diagram:

https://github.com/katzenpost/docs/blob/mix_pipeline_desc.0/HACKING/index.rst#provider-and-mix-pipeline-diagram

Furthermore note that I have provided a sample "echo" services written in golang, python and rust:
https://github.com/katzenpost/server_plugins

This feature adds an additional field to the Provider section of the configuration; basically instead of adding a list of Kaetzchen you can add a list of PluginKaetzchen which specify the command to execute and the desired concurrency level on a per command basis.

**Comment by david415** [comment]
how the fuck am i supposed to fix the dot travis yml file?

**Comment by nogoegst** [comment]
@david415, afaik `strings.Builder` is new for Go 1.9 or something.
The build breaks here: `internal/provider/kaetzchen/plugins.go:173: undefined: strings.Builder`
So either drop Go 1.8 from `travis.yml` (better) or use `bytes.Buffer`.

**Comment by mixmasala** [comment]
Steps followed to test this branch:
checked out this branch (kaetzchen_plugin_system.2)
checked out daemons server_plugins.0.add_parameters.2
cd server/plugin/proto && protoc --go_out=. *.proto
cd ../../../daemons/server && go build

returns:

# github.com/katzenpost/server/plugin
../../server/plugin/grpc.go:51:9: undefined: "github.com/katzenpost/server/plugin/proto".KaetzchenClient

Weird. So I thought I'd check that I was using the right branches using dep ensure; but it looks like Gopkg.lock has the server repo branch pinned to kaetzchen_plugin_system.0.add_parameters.1.

I tried using protoc with a few other branches (kaetzchen_plugin_system.0.add_parameters.1, kaetzchen_plugin_system.0.add_parameters.2) and similarly see the undefined error. What am I doing wrong?

**Comment by mixmasala** [comment]
Aha, I see that I was doing it wrong and needed to tell protoc to use the grpc plugin. Duh.

**Comment by mixmasala** [comment]
Works for me using the branches mentioned above - I tested the echo-go service on a local testnet with cliclient. The new Config options work as expected as well.

File: README.rst [hunk 1]
```diff
@@ -14,6 +14,27 @@ To build the server see the 'daemons' repo:
...
 
 
+Building
+========
+
+* Note that this library requires a ``go generate`` step before running
+``go build`` in order to generate the protobuf and gRPC code for the
+Kaetzchen service plugin system. This means you must have some protobuf
+related deps installed such as:
+
```
Comment (mixmasala) [inline README.rst]: It seems that the "go generate" step needs to happen from the root of this repository before using "go build" in the daemons/server repository.

**Pull request merged**